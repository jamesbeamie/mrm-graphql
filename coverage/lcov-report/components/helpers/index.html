<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for components/helpers</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> components/helpers
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">12.12% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>16/132</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">6.35% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>4/63</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">10% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>2/20</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">12.6% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>16/127</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Controls.js"><a href="Controls.js.html">Controls.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

<tr>
	<td class="file low" data-value="ImageUpload.js"><a href="ImageUpload.js.html">ImageUpload.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	</tr>

<tr>
	<td class="file low" data-value="InputValidators.js"><a href="InputValidators.js.html">InputValidators.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="32" class="abs low">0/32</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	</tr>

<tr>
	<td class="file low" data-value="NavLink.js"><a href="NavLink.js.html">NavLink.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="15" class="abs low">0/15</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	</tr>

<tr>
	<td class="file low" data-value="QueriesHelpers.js"><a href="QueriesHelpers.js.html">QueriesHelpers.js</a></td>
	<td data-value="30.3" class="pic low"><div class="chart"><div class="cover-fill" style="width: 30%;"></div><div class="cover-empty" style="width:70%;"></div></div></td>
	<td data-value="30.3" class="pct low">30.3%</td>
	<td data-value="33" class="abs low">10/33</td>
	<td data-value="44.44" class="pct low">44.44%</td>
	<td data-value="9" class="abs low">4/9</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="5" class="abs low">1/5</td>
	<td data-value="30.3" class="pct low">30.3%</td>
	<td data-value="33" class="abs low">10/33</td>
	</tr>

<tr>
	<td class="file low" data-value="RangePicker.js"><a href="RangePicker.js.html">RangePicker.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="23" class="abs low">0/23</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="23" class="abs low">0/23</td>
	</tr>

<tr>
	<td class="file low" data-value="TableBody.js"><a href="TableBody.js.html">TableBody.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	</tr>

<tr>
	<td class="file medium" data-value="TableHead.js"><a href="TableHead.js.html">TableHead.js</a></td>
	<td data-value="50" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 50%;"></div><div class="cover-empty" style="width:50%;"></div></div></td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="4" class="abs medium">2/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="4" class="abs medium">2/4</td>
	</tr>

<tr>
	<td class="file low" data-value="timeConverter.js"><a href="timeConverter.js.html">timeConverter.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Thu Apr 11 2019 19:58:40 GMT+0300 (East Africa Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
