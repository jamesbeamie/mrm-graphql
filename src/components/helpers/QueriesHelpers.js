import { GET_USER_QUERY, GET_LOCATION_QUERY, GET_ROLE_QUERY } from '../../graphql/queries/People';
import {
  GET_ALL_ROOMS,
  GET_LOCATIONS_QUERY,
  GET_ROOMS_QUERY,
  GET_ALL_REMOTE_ROOMS,
  GET_REMOTE_ROOMS_ALL_LOCATIONS,
} from '../../graphql/queries/Rooms';
import { GET_RESOURCES_QUERY, GET_ROOM_RESOURCES } from '../../graphql/queries/Resources';
import SINGLE_ROOM_FEEDBACK, { ALL_ROOM_FEEDBACK } from '../../graphql/queries/RoomFeedback';
import ALL_ANALYTICS, { ANALYTICS_FOR_DAILY_ROOM_EVENTS } from '../../graphql/queries/analytics';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import apolloClient from '../../utils/ApolloClient';
import GET_ALL_LEVELS from '../../graphql/queries/Levels';
import GET_ROOM_FEEDBACK_QUESTIONS_QUERY from '../../graphql/queries/questions';
import { GET_DEVICES_QUERY } from '../../graphql/queries/Devices';

const getUserLocation = (client = apolloClient) => client.readQuery({
  query: GET_LOCATION_QUERY,
}).userLocation;

const getUserDetails = async (client = apolloClient) => {
  const { UserInfo: userData } = decodeTokenAndGetUserData() || {};
  const email = process.env.NODE_ENV === 'test' ? 'converge@andela.com' : userData.email;
  try {
    const data = client.readQuery(
      {
        query: GET_USER_QUERY,
        variables: {
          email,
        },
      },
      true,
    );
    const user = Object.assign({}, data.user);
    user.firstName = userData.firstName;
    return user;
  } catch (err) {
    const { data } = await client.query({
      query: GET_USER_QUERY,
      variables: { email },
    });
    const user = Object.assign({}, data.user);
    user.firstName = userData.firstName;
    return user;
  }
};

const getAllLocations = async (client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: GET_LOCATIONS_QUERY,
      },
      true,
    );
    return data.allLocations;
  } catch (err) {
    const { data } = await client.query({
      query: GET_LOCATIONS_QUERY,
    });
    return data.allLocations;
  }
};


const getAllLocationsFromCache = (client = apolloClient) => {
  const data = client.readQuery(
    {
      query: GET_LOCATIONS_QUERY,
    },
    true,
  );
  return data.allLocations;
};

const getRoomListVariables = (userLocation, perPage, page, textVariable) => ({
  location: getUserLocation().name,
  office: '',
  page,
  perPage,
  roomLabels: textVariable,
});

const getRoomList = async (userLocation, perPage, page, textVariable, client = apolloClient) => {
  // making 'textVariable' NULL disables filtering by tags
  const nullTextVariable = '';
  try {
    const data = await client.readQuery({
      query: GET_ROOMS_QUERY,
      variables: getRoomListVariables(userLocation, perPage, page, nullTextVariable),
    });
    return data;
  } catch (error) {
    const { data } = await client.query({
      query: GET_ROOMS_QUERY,
      fetchPolicy: 'network-only',
      variables: getRoomListVariables(userLocation, perPage, page, nullTextVariable),
    });
    return data;
  }
};

const getAllResources = async (perPage, page, client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: GET_RESOURCES_QUERY,
        variables: {
          page: page || 1,
          perPage: perPage || 5,
        },
      },
      true,
    );
    return data.allResources;
  } catch (error) {
    const { data } = await client.query({
      query: GET_RESOURCES_QUERY,
      variables: {
        page: page || 1,
        perPage: perPage || 5,
      },
    });
    return data.allResources;
  }
};

const getRoomResources = async (roomId, client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: GET_ROOM_RESOURCES,
        variables: {
          roomId,
        },
      },
      true,
    );
    return data.getResourcesByRoomId.roomResources;
  } catch (error) {
    const { data } = await client.query({
      query: GET_ROOM_RESOURCES,
      variables: {
        roomId,
      },
    });
    return data.getResourcesByRoomId.roomResources;
  }
};

const getSingleRoomFeedback = async (roomId, client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: SINGLE_ROOM_FEEDBACK,
        variables: {
          roomId,
        },
      },
      true,
    );
    return data.roomResponse;
  } catch (error) {
    const { data } = await client.query({
      query: SINGLE_ROOM_FEEDBACK,
      variables: {
        roomId,
      },
    });
    return data.roomResponse;
  }
};

const getAllRemoteRooms = async (client = apolloClient) => {
  const { data } = await client.query(
    {
      query: GET_ALL_REMOTE_ROOMS,
    },
    true,
  );
  return data.allRemoteRooms;
};

const getRemoteRoomsAllLocations = async (client = apolloClient) => {
  const { data } = await client.query(
    {
      query: GET_REMOTE_ROOMS_ALL_LOCATIONS,
      variables: {
        returnAll: true,
      },
    },
    true,
  );
  return data.allRemoteRooms;
};

const getRoomsStructure = async (client = apolloClient) => {
  try {
    const data = client.readQuery({
      query: GET_ALL_LEVELS,
    }, true);
    return data;
  } catch (error) {
    const { data } = await client.query({
      query: GET_ALL_LEVELS,
    }, true);
    return data;
  }
};

const getRoomFeedbackQuestions = async (client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: GET_ROOM_FEEDBACK_QUESTIONS_QUERY,
      },
      true,
    );
    return data.questions;
  } catch (err) {
    const { data } = await client.query({
      query: GET_ROOM_FEEDBACK_QUESTIONS_QUERY,
    });
    return data.questions;
  }
};

const getAllRooms = async (location, client = apolloClient) => {
  try {
    const data = client.readQuery({
      query: GET_ALL_ROOMS,
      variables: {
        location,
      },
    }, true);
    return data;
  } catch (error) {
    const { data } = await client.query({
      query: GET_ALL_ROOMS,
      variables: {
        location,
      },
    }, true);
    return data;
  }
};

const getAllAnalytics = async (dateValue, client = apolloClient) => {
  const variables = {
    ...dateValue,
    locationId: Number(getUserLocation().id),
  };
  try {
    const data = client.readQuery(
      {
        query: ALL_ANALYTICS,
        variables,
      },
      true,
    );
    return data;
  } catch (err) {
    const { data } = await client.query({
      query: ALL_ANALYTICS,
      variables,
    });
    return data;
  }
};

const query = (perPage, page, dateValue) => {
  const { startDate, endDate } = dateValue;
  return ({
    query: ANALYTICS_FOR_DAILY_ROOM_EVENTS,
    variables: {
      startDate,
      endDate,
      page: page || 1,
      perPage: perPage || 5,
    },
  });
};

const getAnalyticForDailyRoomsEvents = async (dateValue, perPage, page, client = apolloClient) => {
  try {
    const data = client.readQuery(
      query(perPage, page, dateValue),
      true,
    );
    return data;
  } catch (err) {
    try {
      const res = await client.query(
        query(perPage, page, dateValue),
      );
      return res.data;
    } catch (error) {
      return error.message;
    }
  }
};

const getAllDevices = async (client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: GET_DEVICES_QUERY,
      },
      true,
    );
    return data.allDevices;
  } catch (error) {
    const { data } = await client.query({
      query: GET_DEVICES_QUERY,
    });
    return data.allDevices;
  }
};

export const getAllResponses = async ({ startDate, endDate } = {}, client = apolloClient) => {
  try {
    const data = client.readQuery(
      {
        query: ALL_ROOM_FEEDBACK,
        variables: {
          startDate: startDate || '',
          endDate: endDate || '',
        },
      },
      true,
    );
    return data.allRoomResponses;
  } catch (error) {
    const { data } = await client.query({
      query: ALL_ROOM_FEEDBACK,
      variables: {
        startDate: startDate || '',
        endDate: endDate || '',
      },
    });
    return data.allRoomResponses;
  }
};

const getUserRole = (client = apolloClient) => client.readQuery({
  query: GET_ROLE_QUERY,
}).userRole;

export {
  getRoomResources,
  getAnalyticForDailyRoomsEvents,
  getAllAnalytics,
  getAllRooms,
  getUserDetails,
  getAllLocations,
  getAllResources,
  getAllRemoteRooms,
  getRoomList,
  getRoomsStructure,
  getRoomFeedbackQuestions,
  getRemoteRoomsAllLocations,
  getAllDevices,
  getSingleRoomFeedback,
  getUserLocation,
  getUserRole,
  getAllLocationsFromCache,
};
