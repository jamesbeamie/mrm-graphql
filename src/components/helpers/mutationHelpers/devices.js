/* eslint-disable import/prefer-default-export */
import apolloClient from '../../../utils/ApolloClient';
import { ADD_DEVICE_MUTATION, EDIT_DEVICE_MUTATION, DELETE_DEVICE_MUTATION } from '../../../graphql/mutations/Devices';
import { GET_DEVICES_QUERY } from '../../../graphql/queries/Devices';
import { GET_ROOMS_QUERY } from '../../../graphql/queries/Rooms';

export const addDeviceMutation = async ({
  name, roomId, deviceType, location,
}, client = apolloClient) => {
  await client
    .mutate({
      mutation: ADD_DEVICE_MUTATION,
      name: 'addDeviceMutation',
      variables: {
        name, deviceType, roomId,
      },
      // write and read from cache
      update: async (proxy, { data: { createDevice: { device } } }) => {
        const cachedDevices = proxy.readQuery({
          query: GET_DEVICES_QUERY,
        });
        const deviceList = cachedDevices.allDevices;
        cachedDevices.allDevices = [...deviceList, device];

        proxy.writeQuery({
          query: GET_DEVICES_QUERY,
          data: cachedDevices,
        });

        const cachedRooms = proxy.readQuery({
          query: GET_ROOMS_QUERY,
          variables: {
            page: 1, perPage: 8, location, office: '', roomLabels: '',
          },
        });

        const target = cachedRooms.allRooms.rooms.findIndex(({ id }) => id === roomId);

        const roomsDeviceList = cachedRooms.allRooms.rooms[target].devices;

        cachedRooms.allRooms.rooms[target].devices = [...roomsDeviceList, device];

        proxy.writeQuery({
          query: GET_ROOMS_QUERY,
          variables: {
            location,
            office: '',
            page: 1,
            perPage: 8,
            roomLabels: '',
          },
          data: cachedRooms,
        });
      },
    });
};

export const editDeviceMutation = async (variables, client = apolloClient) => {
  // edit device mutation
  await client.mutate({
    name: 'editDeviceMutation',
    mutation: EDIT_DEVICE_MUTATION,
    variables,
    // write and read from cache
    update: async (proxy, { data: { updateDevice: { device } } }) => {
      const cachedDevices = proxy.readQuery({ query: GET_DEVICES_QUERY });
      const deviceList = cachedDevices.allDevices;
      deviceList.forEach(({ id }, i) => { // update item in cache
        if (id === device.id) { cachedDevices[i] = device; }
      });

      proxy.writeQuery({
        query: GET_DEVICES_QUERY,
        data: cachedDevices,
      });
    },
  });
};

const newDeviceslist = (devicesList, device) => {
  const list = [];
  devicesList.forEach(({ id }, i) => {
    if (id !== device.id) { list.push(devicesList[i]); }
  });
  return list;
};

const updatedCachedRooms = (rooms, target, device) => {
  const targetRoom = target;
  const otherRooms = rooms.filter((room) => {
    const result = room.id !== device.roomId;
    return result;
  });
  targetRoom.devices = newDeviceslist(targetRoom.devices, device);
  otherRooms.push(targetRoom);
  return otherRooms;
};

const cachedRoomslist = (location, proxy) => proxy.readQuery({
  // cachedRooms read from cache
  query: GET_ROOMS_QUERY,
  variables: {
    page: 1,
    perPage: 8,
    location,
    office: '',
    roomLabels: '',
  },
});

const targetRoom = (rooms, device) => rooms.filter((room) => {
  const result = parseInt(room.id, 10) === device.roomId;
  return result;
})[0];

export const deleteDeviceMutation = async (variables, client = apolloClient) => {
  await client
    .mutate({
      mutation: DELETE_DEVICE_MUTATION,
      name: 'deleteDeviceMutation',
      variables, // write and read from cache
      update: async (proxy, { data: { deleteDevice: { device } } }) => {
        try { // write and read from cache
          const cachedDevices = proxy.readQuery({ query: GET_DEVICES_QUERY });
          cachedDevices.allDevices = newDeviceslist(cachedDevices.allDevices, device);
          proxy.writeQuery({ query: GET_DEVICES_QUERY, data: cachedDevices });

          const { location } = variables;
          const cachedRooms = cachedRoomslist(location, proxy);
          const target = targetRoom(cachedRooms.allRooms.rooms, device);

          cachedRooms.allRooms.rooms =
            updatedCachedRooms(cachedRooms.allRooms.rooms, target, device);

          proxy.writeQuery({
            query: GET_ROOMS_QUERY,
            variables: {
              location, office: '', page: 1, perPage: 8, roomLabels: '',
            },
            data: cachedRooms,
          });
        } catch (error) { return null; }
        return null;
      },
    });
};
