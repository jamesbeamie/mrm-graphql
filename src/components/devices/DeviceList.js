import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import '../../assets/styles/devicelist.scss';
import TableHead from '../helpers/TableHead';
import Device from './Device';
import dummyDevices from '../../fixtures/devices';
import { getAllDevices, getAllRooms } from '../helpers/QueriesHelpers';
import Overlay from '../commons/Overlay';
import ErrorIcon from '../commons/ErrorIcon';
import Pagination from '../commons/Pagination';
import paginate from '../helpers/FrontendPagination';
import DeviceModal from './DeviceModal';

/**
 * Device List Component
 *
 * @returns {JSX}
 */
class DeviceList extends Component {
  state = {
    fetching: true,
    devices: [],
    rooms: [],
    perPage: 5,
    currentPage: 1,
    openModal: false,
    device: {},
  }

  async componentDidMount() {
    this.getData();
  }

  componentDidUpdate(prevProps, { currentPage, perPage }) {
    if (
      currentPage !== this.state.currentPage
      || perPage !== this.state.perPage
    ) {
      this.updatePageDetails();
    }
    if (prevProps.userLocationChanged !== this.props.userLocationChanged) {
      this.toggleFetching();
      setTimeout(() => { this.getData(); }, 1000);
      this.toggleFetching();
    }
  }

  getData = async () => {
    const { location: { name: location, id } } = this.props;
    const devices = await getAllDevices();
    const { allRooms: { rooms: allRooms } } = await getAllRooms(location);
    const rooms = allRooms.filter(({ locationId }) => locationId === Number(id));
    this.setState({
      devices: [...devices].reverse(),
      rooms,
    });
    this.updatePageDetails();
  }

  toggleFetching() { this.setState({ fetching: !this.state.fetching }); }

  handleAction = (action, device) => {
    this.setState({
      openModal: action,
      device,
    });
  }

  deviceComponents = deviceArray =>
    deviceArray.map(device =>
      <Device handleAction={this.handleAction} device={device} key={device.name + device.id} />);


  handlePageChange = (perPageData = 5, currentPageData) => {
    this.setState({
      perPage: Number(perPageData),
      currentPage: Number(currentPageData),
    });
  }

  closeModal = () => this.setState({
    openModal: false,
    modalIsClosed: true,
    device: {},
  });

  updatePageDetails = () => {
    const {
      devices, currentPage, fetching, perPage,
    } = this.state;
    const {
      pageContent: paginatedDeviceList, hasNext, hasPrevious, pages,
    } = paginate(devices,
      { currentPage, perPage },
    );

    this.setState({
      paginatedDeviceList,
      hasNext,
      hasPrevious,
      pages,
      ...(fetching && { fetching: false }),
    });
  }
  render() {
    const {
      device,
      fetching,
      rooms,
      devices,
      currentPage,
      modalIsClosed,
      paginatedDeviceList,
      pages,
      hasNext,
      hasPrevious,
      perPage,
      openModal,
    } = this.state;
    const { location, getRooms } = this.props;

    const displayPaginator = !fetching
      && !(devices.length < 5 && currentPage === 1);

    return (
      <div className="settings-devices-list overlay-container">
        {fetching && <Overlay />}
        <div className="settings-devices-control" />
        <div className="add-new-resource">
          <DeviceModal
            openModal={openModal}
            closeModal={this.closeModal}
            location={location}
            rooms={rooms}
            refetch={this.getData}
            device={device}
            getRooms={getRooms}
            modalIsClosed={modalIsClosed}
          />
        </div>
        {
          !fetching && !devices[0]
            ? (
              <ErrorIcon message="No Device found" />
            ) : (
              <Fragment>
                <div className="table device-table">
                  <TableHead titles={['Name', 'Type', 'Date Added', 'Last Seen', 'Room', 'Location']} />
                  <div className="table__body">{this.deviceComponents((
                    fetching ? dummyDevices : paginatedDeviceList
                  ))}
                  </div>
                </div>
                {
                  displayPaginator &&
                  <div>
                    <Pagination
                      totalPages={pages}
                      hasNext={hasNext}
                      hasPrevious={hasPrevious}
                      handleData={this.handlePageChange}
                      isFetching={fetching}
                      perPage={perPage}
                      currentPage={currentPage}
                    />
                  </div>
                }
              </Fragment>
            )
        }
      </div>
    );
  }
}


DeviceList.propTypes = {
  userLocationChanged: PropTypes.bool,
  location: PropTypes.shape({
    name: PropTypes.string,
  }),
  getRooms: PropTypes.func.isRequired,
};

DeviceList.defaultProps = {
  userLocationChanged: PropTypes.bool,
  location: undefined,
};


export default DeviceList;
