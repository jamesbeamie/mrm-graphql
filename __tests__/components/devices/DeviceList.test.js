import React from 'react';
import { shallow } from 'enzyme';
import DeviceList from '../../../src/components/devices/DeviceList';
import { getAllDevices, getAllRooms } from '../../../src/components/helpers/QueriesHelpers';
import devices from '../../../src/fixtures/devices';

const moreDevices = [{
  name: 'Samsung tablet',
  deviceType: 'Internal Display',
  dateAdded: '14 Aug 2018',
  lastSeen: '16 Feb 2018',
  location: 'Nairobi',
}, {
  name: 'Samsung Galaxy',
  deviceType: 'Internal Display',
  dateAdded: '14 Aug 2018',
  lastSeen: '16 Feb 2018',
  location: 'Nairobi',

}];


jest.mock('../../../src/components/helpers/QueriesHelpers');

describe('DeviceList Component', () => {
  let wrapper;
  beforeAll(() => {
    getAllDevices.mockResolvedValue(devices);
    getAllRooms.mockResolvedValue({
      allRooms: {
        rooms: [
          {
            id: '1230',
            name: 'Another here',
            capacity: 9,
            roomLabels: [
              '4th',
              'Kampala',
            ],
            roomType: 'Meeting room',
            imageUrl: 'https://firebasestorage.googleapis.com/v0/b/andelamrm.appspot.com/o/upload%2F60410220_840773989622302_6083125467391459328_n.jpg?alt=media&token=a9c3cade-8c6a-4ee5-81ea-436e530364f2',
            structureId: '8261c030-7ae9-4091-afb9-a2f4a457843f',
            locationId: 2,
          },
        ],
      },
    });
    wrapper = shallow(<DeviceList
      location={{
        name: 'Lagos',
        id: 2,
      }}
      userLocationChanged={false}
    />);
  });
  it('renders correctly from memory', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('updatePageDetails behaves as expected', () => {
    wrapper.instance().updatePageDetails();
    expect(wrapper.state('hasNext')).toBe(false);
  });

  it('handles page change correctly', () => {
    const mock = jest.spyOn(wrapper.instance(), 'updatePageDetails');
    wrapper.instance().handlePageChange(4, 1);
    wrapper.instance().handlePageChange(undefined, 1);
    expect(wrapper.state('currentPage')).toBe(1);
    expect(mock).toHaveBeenCalled();
  });

  it('getData method works as expected', async () => {
    await wrapper.instance().getData();
    expect(getAllDevices).toHaveBeenCalled();
    expect(getAllRooms).toHaveBeenCalled();
  });

  it('renders error icon', async () => {
    getAllDevices.mockResolvedValue([]);
    wrapper = shallow(<DeviceList location={{
      name: 'Lagos',
      id: 2,
    }}
    />);
    await wrapper.instance().getData();
    expect(wrapper.exists('ErrorIcon')).toBeTruthy();
  });

  it('renders pagination component', async () => {
    getAllDevices.mockResolvedValue(devices.concat(moreDevices));
    wrapper = shallow(<DeviceList
      location={{
        name: 'Lagos',
        id: 2,
      }}
    />);
    await wrapper.instance().getData();
    expect(wrapper.exists('Pagination')).toBeTruthy();
  });

  it('calls handle action', async () => {
    wrapper.instance().handleAction('edit', {});
    expect(wrapper.state('openModal')).toBe('edit');
  });

  it('calls closeModal', async () => {
    wrapper.instance().closeModal();
    expect(wrapper.state('openModal')).toBe(false);
  });

  it('calls componentDidUpdate', async () => {
    wrapper.instance().componentDidUpdate({ userLocationChanged: true }, { a: 1, b: 5 });
    expect(getAllDevices).toHaveBeenCalled();
  });
});
