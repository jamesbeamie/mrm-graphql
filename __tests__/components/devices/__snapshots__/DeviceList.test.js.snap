// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceList Component renders correctly from memory 1`] = `
<div
  className="settings-devices-list overlay-container"
>
  <div
    className="settings-devices-control"
  />
  <div
    className="add-new-resource"
  >
    <DeviceFormModal
      closeModal={[Function]}
      device={Object {}}
      location={
        Object {
          "id": 2,
          "name": "Lagos",
        }
      }
      openModal={false}
      refetch={[Function]}
      rooms={
        Array [
          Object {
            "capacity": 9,
            "id": "1230",
            "imageUrl": "https://firebasestorage.googleapis.com/v0/b/andelamrm.appspot.com/o/upload%2F60410220_840773989622302_6083125467391459328_n.jpg?alt=media&token=a9c3cade-8c6a-4ee5-81ea-436e530364f2",
            "locationId": 2,
            "name": "Another here",
            "roomLabels": Array [
              "4th",
              "Kampala",
            ],
            "roomType": "Meeting room",
            "structureId": "8261c030-7ae9-4091-afb9-a2f4a457843f",
          },
        ]
      }
    />
  </div>
  <div
    className="table device-table"
  >
    <TableHead
      titles={
        Array [
          "Name",
          "Type",
          "Date Added",
          "Last Seen",
          "Room",
          "Location",
        ]
      }
    />
    <div
      className="table__body"
    >
      <Device
        device={
          Object {
            "dateAdded": "14 Aug 2018",
            "deviceType": "Internal Display",
            "lastSeen": "16 Feb 2018",
            "location": "Nairobi",
            "name": "Samsung Note",
            "room": Object {
              "name": "Akwaba",
            },
          }
        }
        handleAction={[Function]}
        key="Samsung Noteundefined"
      />
      <Device
        device={
          Object {
            "dateAdded": "14 Aug 2018",
            "deviceType": "Internal Display",
            "lastSeen": "16 Feb 2018",
            "location": "Nairobi",
            "name": "Samsung duos",
            "room": Object {
              "name": "Ubuntu",
            },
          }
        }
        handleAction={[Function]}
        key="Samsung duosundefined"
      />
      <Device
        device={
          Object {
            "dateAdded": "14 Aug 2018",
            "deviceType": "Internal Display",
            "lastSeen": "16 Feb 2018",
            "location": "Nairobi",
            "name": "Samsung Galaxy",
            "room": Object {
              "name": "Big Apple",
            },
          }
        }
        handleAction={[Function]}
        key="Samsung Galaxyundefined"
      />
      <Device
        device={
          Object {
            "dateAdded": "14 Aug 2018",
            "deviceType": "Internal Display",
            "lastSeen": "16 Feb 2018",
            "location": "Nairobi",
            "name": "Samsung tablet",
            "room": Object {
              "name": "Gold coast",
            },
          }
        }
        handleAction={[Function]}
        key="Samsung tabletundefined"
      />
    </div>
  </div>
</div>
`;
